libs/core/include/core/mutex.hpp

```diff
@@ -17,15 +17,166 @@
+  void QueueUpFor(DebugMutex *mutex, std::thread::id thread, LockLocation const &location);
+  void DeadlockDetected(std::string const &message)
+  {
+    if (static_cast<bool>(throw_on_deadlock_))
+    {
+      throw std::runtime_error(message);
+    }
+
+    std::cerr << message << std::endl;
+    abort();
+  }
+
+  static void ThrowOnDeadlock()
+  {
+    throw_on_deadlock_ = true;
+  }
+
+  static void AbortOnDeadlock()
+  {
+    throw_on_deadlock_ = false;
+  }
+
+private:
+  void FindDeadlock(DebugMutex *first_mutex, std::thread::id thread, LockLocation const &location);
+  std::string CreateTrace(DebugMutex *first_mutex, std::thread::id thread,
+                          LockLocation const &location);
+
+  static std::atomic<bool>                          throw_on_deadlock_;
+  std::mutex                                        mutex_;
+  std::unordered_map<DebugMutex *, std::thread::id> lock_owners_;
+  std::unordered_map<std::thread::id, DebugMutex *> waiting_for_;
+
+  std::unordered_map<DebugMutex *, LockLocation>    lock_location_;
+  std::unordered_map<std::thread::id, LockLocation> waiting_location_;
+};
+
+void RegisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread,
+                              LockLocation const &location = LockLocation());
+void UnregisterMutexAcquisition(DebugMutex *mutex, std::thread::id thread);
+void QueueUpFor(DebugMutex *mutex, std::thread::id thread,
+                LockLocation const &location = LockLocation());
+
+class DebugMutex : public std::mutex
+{
+public:
+  DebugMutex()  = default;
+  ~DebugMutex() = default;
+
+  void lock(std::string const &filename, int32_t line)
+  {
+    LockLocation loc{filename, line};
+    QueueUpFor(this, std::this_thread::get_id(), loc);
+    std::mutex::lock();
+    RegisterMutexAcquisition(this, std::this_thread::get_id(), loc);
+  }
+
+  void lock()
+  {
+    QueueUpFor(this, std::this_thread::get_id());
+    std::mutex::lock();
+    RegisterMutexAcquisition(this, std::this_thread::get_id());
+  }
+
+  void unlock()
+  {
+    UnregisterMutexAcquisition(this, std::this_thread::get_id());
+    std::mutex::unlock();
+  }
+
+  bool try_lock()
+  {
+    if (std::mutex::try_lock())
+    {
+      RegisterMutexAcquisition(this, std::this_thread::get_id());
+      return true;
+    }
+
+    return false;
+  }
+};
+
+template <typename T>
+class DebugLockGuard
+{
+public:
+  DebugLockGuard(T &lockable, std::string const & /*filename*/, int32_t /*line*/)
+    : lockable_{lockable}
+  {
+    lockable_.lock();
+  }
+
+  ~DebugLockGuard()
+  {
+    lockable_.unlock();
+  }
+
+private:
+  T &lockable_;
+};
+
+template <>
+class DebugLockGuard<DebugMutex>
+{
+public:
+  DebugLockGuard(DebugMutex &lockable, std::string const &filename, int32_t line)
+    : lockable_{lockable}
+  {
+    lockable_.lock(filename, line);
+  }
+
+  ~DebugLockGuard()
+  {
+    lockable_.unlock();
+  }
+
+private:
+  DebugMutex &lockable_;
+};
+
+//#ifndef NDEBUG
+using Mutex = DebugMutex;
```

Comment:
Can we control this via the existing cmake option instead of as default in debug mode please